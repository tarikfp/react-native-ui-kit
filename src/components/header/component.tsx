import PropTypes from "prop-types";
import React from "react";
import {
  ImageStyle,
  StyleProp,
  StyleSheet,
  TextStyle,
  View,
  ViewProps,
  ViewStyle,
} from "react-native";
import { BaseColors, BaseFonts, BaseShadow } from "../../theme";
import { StylePropType } from "../utils";
import HeaderIcon from "./components/icon";
import HeaderImage from "./components/image";
import HeaderTitle from "./components/title";

const DEFAULT_HEADER_HEIGHT = 56;

export type Props = ViewProps & {
  /**
   * Style of the header
   */
  readonly style?: StyleProp<ViewStyle>;
  /**
   * List item height
   * @default 56
   */
  readonly height?: number;
  /**
   * Applies to shadow to the list item.
   * @default false
   */
  readonly hasShadow?: boolean;
  /**
   * Background color of the header.
   * @default #0075E1
   */
  readonly backgroundColor?: string;
};

/**
 * ## Usage
 * ```js
 *     <Header height={72}>
 *       <Header.Title label="Home screen" />
 *       <Header.Image
 *         source={{
 *           uri: "https://reactnative.dev/img/logo-og.png",
 *         }}
 *       />
 *       <Header.Icon
 *         alignment="end"
 *         name="arrow-right"
 *         type="MaterialCommunityIcons"
 *       />
 *     </Header>
 * ```
 */

function Header({
  height = DEFAULT_HEADER_HEIGHT,
  hasShadow = false,
  style,
  backgroundColor = "#0075E1",
  children,
  ...touchableOpacityProps
}: Props) {
  const shadowStyle = hasShadow ? BaseShadow : undefined;

  return (
    <View
      style={StyleSheet.flatten<ViewStyle>([
        styles.wrapper,
        { height, backgroundColor },
        shadowStyle,
        style,
      ])}
      {...touchableOpacityProps}>
      {React.Children.toArray(children)
        .filter((child) => child !== null && typeof child !== "boolean")
        .map((child, childInd) => {
          if (!React.isValidElement(child)) {
            return null;
          }

          let props:
            | {
                style?: StyleProp<ViewStyle> | StyleProp<TextStyle>;
                wrapperStyle?: StyleProp<ViewStyle>;
              }
            | undefined;

          if (child.type === HeaderTitle) {
            props = {
              wrapperStyle: {
                height,
              },
              style: [
                BaseFonts.h3,
                {
                  color: BaseColors.textWhite,
                  marginLeft: childInd !== 0 ? 8 : 0,
                },
                child.props.style,
              ],
            };
          }

          if (child.type === HeaderIcon) {
            props = {
              style: [styles.icon, child.props.style],
            };
          }

          if (child.type === HeaderImage) {
            props = {
              style: [styles.image, child.props.style],
            };
          }

          return React.cloneElement(child, props);
        })}
    </View>
  );
}

Header.Icon = HeaderIcon;
Header.Image = HeaderImage;
Header.Title = HeaderTitle;

export default Header;

Header.propTypes = {
  style: StylePropType,
  height: PropTypes.number,
  hasShadow: PropTypes.bool,
  backgroundColor: PropTypes.string,
};

interface Styles {
  wrapper: ViewStyle;
  icon: TextStyle;
  image: ImageStyle;
}

const styles = StyleSheet.create<Styles>({
  wrapper: {
    flexDirection: "row",
    width: "100%",
    alignItems: "center",
    paddingHorizontal: 16,
    paddingVertical: 8,
    borderWidth: StyleSheet.hairlineWidth,
    height: 56,
    borderColor: BaseColors.borderColor,
    backgroundColor: BaseColors.paper,
  },
  icon: {
    fontSize: 26,
    marginHorizontal: 4,
  },
  image: {
    marginHorizontal: 4,
    width: 30,
    height: 36,
  },
});
