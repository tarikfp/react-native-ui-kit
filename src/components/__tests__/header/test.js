import * as React from "react";
import renderer from "react-test-renderer";
import { Header } from "../../header";

describe("tests header component", () => {
  it("renders header with title", () => {
    const tree = renderer
      .create(
        <Header>
          <Header.Title title="test" />
        </Header>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders header with icon", () => {
    const tree = renderer
      .create(
        <Header>
          <Header.Icon name="arrow-left" type="MaterialCommunityIcons" />
        </Header>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders header with image", () => {
    const tree = renderer
      .create(
        <Header>
          <Header.Image
            style={{ width: 40, height: 40 }}
            source={{ uri: "https://reactjs.org/logo-og.png" }}
          />
        </Header>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders header with custom height", () => {
    const tree = renderer
      .create(
        <Header height={75}>
          <Header.Title title="Test" />
        </Header>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders header with shadow", () => {
    const tree = renderer
      .create(
        <Header height={75} hasShadow>
          <Header.Title title="Test" />
        </Header>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders header with custom backgroundColor", () => {
    const tree = renderer
      .create(
        <Header height={75} backgroundColor="orange">
          <Header.Title title="Test" />
        </Header>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders header with custom style", () => {
    const tree = renderer
      .create(
        <Header style={{ paddingHorizontal: 20 }}>
          <Header.Title title="Test" />
        </Header>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
