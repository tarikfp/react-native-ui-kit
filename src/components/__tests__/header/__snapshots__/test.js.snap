// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests header component renders header with custom backgroundColor 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "orange",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "500",
            "lineHeight": 30,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "marginLeft": 0,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests header component renders header with custom height 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "500",
            "lineHeight": 30,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "marginLeft": 0,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests header component renders header with custom style 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "paddingHorizontal": 20,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 56,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "500",
            "lineHeight": 30,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "marginLeft": 0,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests header component renders header with icon 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "flex": 0,
          "flexDirection": "row",
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 20,
          },
          Array [
            Object {
              "fontSize": 26,
              "marginHorizontal": 4,
            },
            undefined,
          ],
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰁍
    </Text>
  </View>
</View>
`;

exports[`tests header component renders header with image 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://reactjs.org/logo-og.png",
      }
    }
    style={
      Array [
        Object {
          "height": 36,
          "marginHorizontal": 4,
          "width": 30,
        },
        Object {
          "height": 40,
          "width": 40,
        },
      ]
    }
  />
</View>
`;

exports[`tests header component renders header with shadow 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 75,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "shadowColor": "rgba(45, 97, 185, 0.04)",
      "shadowOffset": Object {
        "height": 4,
        "width": 3,
      },
      "shadowOpacity": 1.25,
      "shadowRadius": 1,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 75,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "500",
            "lineHeight": 30,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "marginLeft": 0,
            },
            undefined,
          ],
        ]
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests header component renders header with title 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#0075E1",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderWidth": 0.5,
      "flexDirection": "row",
      "height": 56,
      "paddingHorizontal": 16,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "height": 56,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(255, 255, 255, 1)",
            "fontSize": 22,
            "fontWeight": "500",
            "lineHeight": 30,
            "textAlign": "left",
          },
          Array [
            Object {
              "fontSize": 16,
              "fontWeight": "500",
              "lineHeight": 24,
            },
            Object {
              "color": "rgba(255, 255, 255, 1)",
              "marginLeft": 0,
            },
            undefined,
          ],
        ]
      }
    >
      test
    </Text>
  </View>
</View>
`;
